src/parsimmon.js

```diff
@@ -400,7 +400,31 @@ function makeLineColumnIndex(input, i) {
       }
       break;
     }
-    if (input.charAt(j) === "\n") {
+
+    /**
+     * Defined as \r not followed by \n
+     */
+    var isCrAlone = function () {
+      return (input.charAt(j) === "\r" && input.charAt(j + 1) !== "\n");
+    }
+
+    /**
+     * Defined as \n not preceded by \r
+     */
+    var isLfAlone = function () {
+      return (input.charAt(j) === "\n" && input.charAt(j - 1) !== "\r");
+    }
+
+    /**
+     * Defined as \r followed by \n
+     */
+    var isCrLf = function () {
+      return (input.charAt(j) === "\r" && input.charAt(j + 1) === "\n");
+    }
+
+    if (
+      isCrAlone() || isLfAlone() || isCrLf()
+    ) {
```

Comment:
Thinking about this again... is it enough to just check `isCrAlone() || isLfAlone()`?  if you're at a CRLF the CR will just be treated like a normal character for the purpose of line numbers.

Reply:
Also feeling worried like "is this a bug fix or is this a breaking change" haha... I guess if someone was parsing a weird file that has raw carriage returns in it (like a terminal output log), this could mess up their parsing...

Reply:
> Thinking about this again... is it enough to just check isCrAlone() || isLfAlone()?

It won't be enough, a line break can be one of these three things, 

1. a solo LF `\n`
2. a solo CR `\r`
3. a full CRLF `\r\n`

IMO, we have to check, lookback, or lookahead and mark the change of line.

> Also feeling worried like "is this a bug fix or is this a breaking change"

You'll know this best, I believe here though that the only change is the counting of solo CR linebreaks, what's the path forward if you decide this is a breaking change?

Reply:
Worth noting my new version is able to disambiguate between these ...

1. a solo LF `\n`
2. a solo CR `\r`
3. a full CRLF `\r\n`

... correctly since we are checking the next and the previous character appropriately.

Reply:
my reasoning on just checking `\r` without `\n` or just a solo `\n` is that the `\r` in `\r\n` does not need to be treated differently from a regular character. it's still going to not increase the line number, it's still going to increase the column number. so the code should be shorter by just checking those cases

Reply:
I'll be back shortly with a different version, thanks.

Reply:
@wavebeem 

What would help me greatly, i.e. my major blocker while working on this is how to setup the expectations correctly, my sense is that parsing `\n` and `\r\n` should generate similar results yet they don't or at least, not in a way that I understand.

Here's the test again, I am running it on upstream/master and not my change here.

```node
["\n", "\r\n"].forEach(function(eol) {
  it(
    "should correctly report line number when parsing optional " +
      JSON.stringify(eol),
    function() {
      var parser = Parsimmon.seq(
        Parsimmon.newline,
        Parsimmon.newline.many()
      ).mark();

      assert.deepEqual(parser.parse(eol).value, {
        start: {
          column: 0,
          line: 0,
          offset: 0
        },
        end: {
          column: 0,
          line: 0,
          offset: 0
        },
        value: [eol, []]
      });
    }
  );
});
```

And here's the output

```diff
  1)  should correctly report line number when parsing optional "\n":

      AssertionError: expected { Object (start, value, ...) } to deeply equal { Object (start, end, ...) }
      + expected - actual

       {
         "end": {
      -    "column": 1
      -    "line": 2
      -    "offset": 1
      +    "column": 0
      +    "line": 0
      +    "offset": 0
         }
         "start": {
           "column": 0
      -    "line": 2
      +    "line": 0
           "offset": 0
         }
         "value": [
           "\n"

      at Context.<anonymous> (test/core/mark.test.js:33:14)
      at processImmediate (internal/timers.js:461:21)

  2)  should correctly report line number when parsing optional "\r\n":

      AssertionError: expected { Object (start, value, ...) } to deeply equal { Object (start, end, ...) }
      + expected - actual

       {
         "end": {
      -    "column": 1
      -    "line": 2
      -    "offset": 2
      +    "column": 0
      +    "line": 0
      +    "offset": 0
         }
         "start": {
      -    "column": 1
      -    "line": 1
      +    "column": 0
      +    "line": 0
           "offset": 0
         }
         "value": [
           "\r\n"
```

Why in the case of parsing `\n` only do we expect this result?

```
{
  start: {
    column: 0,
    line: 2,
    offset: 0
  },
  end: {
    column: 1,
    line: 2,
    offset: 1
  }
}
```

And when parsing `\r\n` we expect this result?

```
{
  start: {
    column: 1,
    line: 1,
    offset: 0
  },
  end: {
    column: 1,
    line: 2,
    offset: 2
  }
}
```

I won't be able to introduce this fix until I understand how this works and I don't see a reason for the `start.line` and `start.column` for example to be so different between `\n` and `\r\n`.

What am I missing?

Reply:
Check this commit for the latest version, I updated the test setup to use a line of text and iterating over the different line breaks: https://github.com/jneen/parsimmon/pull/329/commits/31311427ac3323c632b9d7ac104e946dc7415805

I'll delete my last commit I think I started from the wrong parser setup to exercise this.

Reply:
> Also feeling worried like "is this a bug fix or is this a breaking change"

I've been thinking more about this. Considering the documentation does not state what is considered a newline for the purpose of `P.index`, either answer seems reasonable. But considering that `P.newline` works for CR, LF, or CRLF line endings, I think it's reasonable to assume that Parsimmon supports ancient CR only text files.

So I think this is truly a bugfix and not a breaking change.

Reply:
@wavebeem have you checked the latest version? are you good with the change, I feel with the updated test setup that this change is correct.

Reply:
I'll have to look later when I have more time, just wanted to let you know my thoughts on the bug vs breaking status :)